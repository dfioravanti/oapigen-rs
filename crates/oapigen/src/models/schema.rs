use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use std::hash::Hash;
use std::{collections::HashMap, fmt};

/// Collects all the imports that were already added to the code.
/// It is used to determine if an import should be added or not.
/// For example assume that two different structs need
/// chrono::DateTime then we add "chrono::DateTime" to the Imports
/// for the first struct that we hit and we will not add it in the second case since we already have it.
pub type Imports = HashMap<String, TokenStream>;

#[derive(Debug)]
pub enum CurrentType {
    Type,
    Const,
    Enum,
    Vec,
    Struct,
}

/// SchemaAsRust represents a rust type that was inferred from a OpenAPI specification.
/// It contains all the information needed to render the type correctly.
#[derive(Debug)]
pub struct SchemaAsRust {
    /// The name of the field as a TokenStream
    pub tokenized_name: TokenStream,
    /// The type represented as a TokenStream
    pub tokenized_type: TokenStream,
    /// The imports needed to make the type compile.
    pub imports: Imports,

    pub current_type: CurrentType,
}
impl Eq for SchemaAsRust {}
impl PartialEq for SchemaAsRust {
    fn eq(&self, other: &Self) -> bool {
        self.to_token_stream().to_string() == other.to_token_stream().to_string()
    }
}

impl Hash for SchemaAsRust {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.to_token_stream().to_string().hash(state);
    }
}

impl fmt::Display for SchemaAsRust {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let imports = self
            .imports
            .iter()
            .fold(TokenStream::new(), |mut acc, (_, v)| {
                acc.extend(v.clone());
                acc
            });

        if imports.is_empty() {
            write!(f, "{}", self.to_token_stream())
        } else {
            write!(f, "{} \n {}", imports, self.to_token_stream())
        }
    }
}

impl ToTokens for SchemaAsRust {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let generated_tokens = match self.current_type {
            CurrentType::Type => tokenize_type(self),
            CurrentType::Const => todo!(),
            CurrentType::Enum => todo!(),
            CurrentType::Vec => todo!(),
            CurrentType::Struct => todo!(),
        };

        tokens.clone_from(&generated_tokens);
    }
}

fn tokenize_type(tokenized_schema: &SchemaAsRust) -> TokenStream {
    let tokenized_name = &tokenized_schema.tokenized_name;
    let tokenized_type = &tokenized_schema.tokenized_type;

    quote! { struct #tokenized_name(#tokenized_type); }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    #[test]
    fn test_parsed_schema_display() {
        let tokenized_schema = quote! {
            DateTime
        };

        let mut imports = HashMap::new();
        imports.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime; },
        );

        let parsed_schema = SchemaAsRust {
            tokenized_name: quote! { time },
            tokenized_type: tokenized_schema,
            imports,
            current_type: CurrentType::Type,
        };

        let formatted_parsed_schema = format!("{}", parsed_schema);

        assert_yaml_snapshot!(formatted_parsed_schema);
    }
}
