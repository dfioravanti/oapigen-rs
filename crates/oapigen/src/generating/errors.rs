use std::io;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum GeneratingError {
    #[error("could not write to file.")]
    WriteToFileError(#[from] io::Error),
}
