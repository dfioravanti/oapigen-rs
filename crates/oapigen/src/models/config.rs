use std::path::PathBuf;

pub enum DateTimeLibraries {
    Chrono,
    Jiff,
}

pub struct Libraries {
    pub datetime: DateTimeLibraries,
}

impl Default for Libraries {
    fn default() -> Self {
        Libraries {
            datetime: DateTimeLibraries::Chrono,
        }
    }
}

#[derive(Default)]
pub struct Config {
    pub output_path: PathBuf,
    pub libraries: Libraries,
}
