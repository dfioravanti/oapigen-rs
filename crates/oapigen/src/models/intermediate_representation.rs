#[derive(Debug, serde::Serialize)]
pub struct Struct {
    pub name: String,
    pub description: Option<String>,
    pub fields: Vec<Field>,
}

#[derive(Debug, serde::Serialize)]
pub struct Field {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub is_required: bool,
}

pub struct ConstantField {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub value: serde_json::Value,
}
