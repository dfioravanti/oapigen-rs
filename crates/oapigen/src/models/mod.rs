pub mod intermediate_representation;
pub mod schema;

pub use intermediate_representation::*;
pub use schema::*;
