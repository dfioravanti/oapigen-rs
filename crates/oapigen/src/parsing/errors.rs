use oas3::spec;
use proc_macro2::LexError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParsingError {
    #[error("could not parse the string to TokenStream.")]
    CannotParseString(#[from] LexError),
    #[error("could not resolve the reference with the passed spec.")]
    ReferenceNotFound(#[from] spec::RefError),
}
