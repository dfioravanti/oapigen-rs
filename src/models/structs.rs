use proc_macro2::TokenStream;
use quote::{ToTokens, TokenStreamExt, quote};
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, fmt};

/// Collects all the imports that were already added to the code.
/// It is used to determine if an import should be added or not.
/// For example assume that two different structs need
/// chrono::DateTime then we add "chrono::DateTime" to the Imports
/// for the first struct that we hit and we will not add it in the second case since we already have it.
pub type Imports = HashMap<String, TokenStream>;

#[derive(Debug)]
pub enum CurrentType {
    Type,
    Const,
    Enum,
    Vec,
    Struct,
}

/// TokenizedType represents a rust type that was inferred from a OpenAPI specification.
/// It contains all the information needed to render the type correctly.
#[derive(Debug)]
pub struct TokenizedSchema {
    /// The name of the field as a TokenStream
    pub tokenized_name: TokenStream,
    /// The type represented as a TokenStream
    pub tokenized_type: TokenStream,
    /// The imports needed to make the type compile.
    pub imports: Imports,

    pub current_type: CurrentType,
}

impl fmt::Display for TokenizedSchema {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let imports = self
            .imports
            .iter()
            .map(|(_, v)| v.to_string())
            .collect::<Vec<_>>();

        write!(
            f,
            "imports => {:?} \n schema => {:?}",
            imports,
            self.to_token_stream().to_string()
        )
    }
}

impl quote::ToTokens for TokenizedSchema {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let tokenized_name = &self.tokenized_name;
        let tokenized_type = &self.tokenized_type;

        let generated_tokens = quote! {

            #tokenized_name: #tokenized_type

        };

        tokens.clone_from(&generated_tokens);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    #[test]
    fn test_parsed_schema_display() {
        let tokenized_schema = quote! {
            DateTime
        };

        let mut imports = HashMap::new();
        imports.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime },
        );

        let parsed_schema = TokenizedSchema {
            tokenized_name: quote! { time },
            tokenized_type: tokenized_schema,
            imports,
        };

        let formatted_parsed_schema = format!("{}", parsed_schema);

        assert_yaml_snapshot!(formatted_parsed_schema);
    }
}
