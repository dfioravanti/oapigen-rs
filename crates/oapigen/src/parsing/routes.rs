use crate::models;
use crate::models::schema;
use crate::parsing::errors::ParsingError;
use crate::parsing::typeset::tokenize_schema;
use convert_case::{Case, Casing};
use oas3::spec;
use std::option::Option;

pub(crate) fn parse_routes(
    config: &models::Config,
    spec: &oas3::Spec,
) -> Result<Vec<schema::SchemaAsRust>, ParsingError> {
    let mut output = Vec::new();

    let paths = &spec.paths;

    if let Some(paths) = paths {
        for (route_name, path) in paths {
            if path.get.is_some() {
                let parsed =
                    parse_operation(config, spec, &"Get".to_string(), route_name, &path.get)?;
                output.extend(parsed);
            }
        }
    }

    Ok(output)
}

fn parse_operation(
    config: &models::Config,
    spec: &oas3::Spec,
    method_name: &String,
    route_name: &String,
    operation: &Option<spec::Operation>,
) -> Result<Vec<schema::SchemaAsRust>, ParsingError> {
    let mut output = Vec::new();
    if let Some(operation) = operation {
        let operation_name = match &operation.operation_id {
            Some(operation_id) => operation_id.to_case(Case::UpperCamel),
            None => [method_name, &route_name.to_case(Case::UpperCamel)]
                .map(String::as_str)
                .concat(),
        };

        let responses = &operation.responses;
        if let Some(responses) = responses {
            for (response_name, response) in responses {
                let resolved_response = response.resolve(spec)?;
                let parsed = parse_response(
                    config,
                    spec,
                    &operation_name,
                    &response_name,
                    &resolved_response,
                )?;
                output.extend(parsed);
            }
        }
    }
    Ok(output)
}

fn parse_response(
    config: &models::Config,
    spec: &oas3::Spec,
    operation_name: &String,
    response_name: &String,
    response: &spec::Response,
) -> Result<Vec<schema::SchemaAsRust>, ParsingError> {
    let mut v = Vec::with_capacity(response.content.len());
    for (mediatype_name, mediatype) in &response.content {
        let parsed = parse_mediatype(config, spec, operation_name, response_name, mediatype)?;
        match parsed {
            Some(tokenized_schema) => v.push(tokenized_schema),
            None => {}
        }
    }

    Ok(v)
}

fn parse_mediatype(
    config: &models::Config,
    spec: &oas3::Spec,
    operation_name: &String,
    response_name: &String,
    media_type: &spec::MediaType,
) -> Result<Option<schema::SchemaAsRust>, ParsingError> {
    let schema_name = [
        operation_name.to_case(Case::UpperCamel),
        "Response".to_string(),
        response_name.to_case(Case::UpperCamel),
    ]
    .join("");

    let schema = &media_type.schema;
    if let Some(schema) = schema {
        let parsed_schema = schema.resolve(spec)?;
        let tokens = tokenize_schema(config, schema_name, parsed_schema)?;

        return Ok(Some(tokens));
    }

    Ok(None)
}
