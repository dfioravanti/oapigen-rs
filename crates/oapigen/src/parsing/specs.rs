use crate::models::schema;
use crate::parsing::errors::ParsingError;
use crate::parsing::routes::parse_routes;
use oas3::spec;

pub fn parse_specs(spec: oas3::Spec) -> Result<Vec<schema::SchemaAsRust>, ParsingError> {
    let mut parsed_schemas = Vec::new();

    let components = spec.clone().components;
    if let Some(components) = components {
        let parsed_schemas_from_routes = parse_routes(&spec)?;
        parsed_schemas.extend(parsed_schemas_from_routes);

        for (schema_name, schema) in components.schemas {
            let resolved_schema = schema.resolve(&spec)?;

            let _ = resolved_schema.title.is_none();

            match resolved_schema.schema_type {
                Some(t) => match t {
                    spec::SchemaTypeSet::Single(i) => todo!(),
                    spec::SchemaTypeSet::Multiple(_) => todo!(),
                },
                None => todo!(),
            };
        }

        return Ok(parsed_schemas);
    }

    Ok(Vec::new())
}

#[cfg(test)]
mod tests {
    use crate::parsing::specs::parse_specs;
    use rstest::rstest;

    #[rstest]
    #[case("one route int", "fixtures/one_route_int.yaml")]
    fn test_parse_structs(#[case] name: &str, #[case] path: &str) {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(name);

        let yaml = std::fs::read_to_string(path).unwrap();
        let spec = oas3::from_yaml(yaml).unwrap();

        let got = parse_specs(spec).unwrap();
        let got_as_strings = got.iter().map(|v| v.to_string()).collect::<Vec<_>>();
        settings.bind(|| {
            insta::assert_yaml_snapshot!(got_as_strings);
        });
    }
}
