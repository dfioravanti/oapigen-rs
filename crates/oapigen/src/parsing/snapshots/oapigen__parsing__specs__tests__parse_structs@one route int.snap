---
source: crates/oapigen/src/parsing/specs.rs
expression: got_as_strings
---
- type ListWidgetsResponse200 = i32 ;
